core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/TupleProducer.scala

```diff
@@ -0,0 +1,70 @@
+package edu.uci.ics.amber.engine.architecture.messaginglayer
+
+import edu.uci.ics.amber.engine.architecture.worker.neo.WorkerInternalQueue
+import edu.uci.ics.amber.engine.architecture.worker.neo.WorkerInternalQueue.{
+  EndMarker,
+  EndOfAllMarker,
+  InputTuple,
+  SenderChangeMarker
+}
+import edu.uci.ics.amber.engine.common.ambermessage.WorkerMessage.{DataPayload, EndOfUpstream}
+import edu.uci.ics.amber.engine.common.ambermessage.neo.DataEvent
+import edu.uci.ics.amber.engine.common.ambertag.neo.Identifier
+
+import scala.collection.mutable
+
+class TupleProducer(workerInternalQueue: WorkerInternalQueue) {
```

Comment:
It is not a Producer. It is a Converter. Continue with the BatchToTupleConverter